### PR TITLE
[2.11.x] DDF-3594 improve performance of ldap login 

### PR DESCRIPTION
**Pull request by ryeats**
#### What does this PR do?
Minimizes the number of connections being opened and creates a connection pool per configured LDAP login.
#### Who is reviewing it? 
(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)
#### Select relevant component teams: 
https://github.com/orgs/codice/teams
@derekwilhelm @jlcsmith 
#### Choose 2 committers to review/merge the PR. 
(please choose ONLY two committers from below, delete the rest)
@rzwiefel
@stustison
#### How should this be tested? (List steps with links to updated documentation)
Just run the itests, I ran it with 2 AD LDAPS and with 1 AD LDAPS with my setup that causes ~100 logins at startup but we won't be able to reproduce that with vanilla DDF easily.
#### Any background context you want to provide?
Since we are using the broker and have a bunch of routes with consumers and associated connection pools we cause a lot of LDAP traffic on startup. When LDAPS was configured the overhead from the SSL handshakes and inefficiencies in the LDAP login service caused connections to timeout.
#### What are the relevant tickets?
[DDF-3594](https://codice.atlassian.net/browse/DDF-3594)
#### Screenshots (if appropriate)
connection pools can be monitored with JMX(here I am using hawtio)
screen shot 2018-02-01 at 4 32 37 pm
#### Checklist:
- [ ] Documentation Updated
- [x ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LDAPConnectionPool.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.ldap.ldaplogin;
+
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
+import org.forgerock.opendj.ldap.Connection;
+import org.forgerock.opendj.ldap.LDAPConnectionFactory;
+
+public class LDAPConnectionPool extends GenericObjectPool<Connection> {
```
Comment (dcruver) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LDAPConnectionPool.java]: ✏️ GenericObjectPool is a concrete class, so rather than extending it, can we just wire it up in blueprint?

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/SslLdapLoginModule.java [hunk 2]
```diff
@@ -475,69 +435,6 @@ public boolean login() throws LoginException {
     }
   }
 
-  protected LDAPConnectionFactory createLdapConnectionFactory(String url, Boolean startTls)
```
Comment (ryeats) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/SslLdapLoginModule.java]: moved this all into the LdapLoginConfig class which creates a connection pool and then registers it as a service. This class then manually injects the ConnectionPool. I had to do that because this class is constructed by Class.newInstance() then the initialize method is called however it was called more than once so I was getting ~70 connection pools per LDAP instead of 1.

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/SslLdapLoginModule.java [hunk 3]
```diff
@@ -438,19 +385,32 @@ public void initialize(
     roleFilter = (String) options.get(ROLE_FILTER);
     roleNameAttribute = (String) options.get(ROLE_NAME_ATTRIBUTE);
     roleSearchSubtree = Boolean.parseBoolean((String) options.get(ROLE_SEARCH_SUBTREE));
-    startTls = Boolean.parseBoolean(String.valueOf(options.get(SSL_STARTTLS)));
     setBindMethod((String) options.get(BIND_METHOD));
     realm = (String) options.get(REALM);
     kdcAddress = (String) options.get(KDC_ADDRESS);
+    String connectionPoolId = (String) options.get(CONNECTION_POOL_ID);
+    installLdapConnectionPool(connectionPoolId);
```
Comment (ryeats) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/SslLdapLoginModule.java]: We could simplify the SslLdapLoginModule class and not have to pass as many properties around by moving all the logic to LdapLoginConfig and just pass a service reference to this class and call the corresponding methods required. May save that for another ticket though if people like the idea...

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapConnectionPooledObjectFactory.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  @Override
+  public PooledObject<Connection> wrap(Connection connection) {
+    return new DefaultPooledObject<>(connection);
+  }
+
+  @Override
+  public void destroyObject(PooledObject<Connection> p) throws Exception {
+    if (p != null && p.getObject() != null && !p.getObject().isClosed()) {
```
[the remark this answers is not in the document]
Comment (ryeats) [reply platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapConnectionPooledObjectFactory.java]: hmmm p.getObject().isValid() could return false and the connection should still be closed I think but I should probably protect against null anyway

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java [hunk 5]
```diff
@@ -79,12 +103,104 @@
...
+        LDAPConnectionFactory ldapConnectionFactory =
+            createLdapConnectionFactory(
+                (String) ldapProperties.get(LDAP_URL),
+                Boolean.parseBoolean((String) ldapProperties.get(START_TLS)));
+        ldapConnectionPool = new LDAPConnectionPool(ldapConnectionFactory, id);
+      } catch (LdapException e) {
+        LOGGER.error("Error creating ldap connection factory", e);
+      }
+
+      Dictionary<String, String> serviceProps = new Hashtable<>();
```
[the remark this answers is not in the document]
Comment (ryeats) [reply platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java]: ill add a comment to remind myself to do that when I port it to master

**Comment by ryeats** [comment]
build now

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java [hunk 6]
```diff
@@ -79,12 +111,128 @@
...
+            new GenericObjectPool<>(
+                new LdapConnectionPooledObjectFactory(ldapConnectionFactory),
+                createGenericPoolConfig(id),
+                createGenericPoolAbandonConfig());
+
+      } catch (LdapException e) {
+        LOGGER.error("Error creating ldap connection factory", e);
+      }
+
+      // TODO Switch to DictionaryMap.java
```
Comment (jlcsmith) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java]: ❓ remove or fix?

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java [hunk 7]
```diff
@@ -79,12 +111,128 @@
...
+        LDAPConnectionFactory.SSL_ENABLED_CIPHER_SUITES,
+        Arrays.asList(System.getProperty("https.cipherSuites", "").split(",")));
+    lo.set(
+        LDAPConnectionFactory.SSL_ENABLED_PROTOCOLS,
+        Arrays.asList(System.getProperty("https.protocols", "").split(",")));
+    lo.set(
+        LDAPConnectionFactory.TRANSPORT_PROVIDER_CLASS_LOADER,
+        SslLdapLoginModule.class.getClassLoader());
+
+    String host = url.substring(url.indexOf("://") + 3, url.lastIndexOf(':'));
```
Comment (jlcsmith) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java]: ❓ why not use http://docs.oracle.com/javase/7/docs/api/java/net/URI.html for this rather than parsing it yourself?

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java [hunk 8]
```diff
@@ -79,12 +111,128 @@
...
+    }
+
+    lo.set(LDAPConnectionFactory.HEARTBEAT_TIMEOUT, new Duration(30L, TimeUnit.SECONDS));
+    lo.set(LDAPConnectionFactory.HEARTBEAT_INTERVAL, new Duration(60L, TimeUnit.SECONDS));
+    lo.set(LDAPConnectionFactory.CONNECT_TIMEOUT, new Duration(30L, TimeUnit.SECONDS));
+
+    lo.set(LDAPConnectionFactory.SSL_USE_STARTTLS, useTls);
+    lo.set(
+        LDAPConnectionFactory.SSL_ENABLED_CIPHER_SUITES,
+        Arrays.asList(System.getProperty("https.cipherSuites", "").split(",")));
```
Comment (jlcsmith) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java]: I believe this is re-used code rather than "new" code, but it is probably a good idea to check that these properties exist before using them in the very unlikely case they aren't set to avoid an NPE.  Also you can use the constants already defined in https://github.com/codice/ddf/blob/master/platform/security/platform-security-core-api/src/main/java/ddf/security/SecurityConstants.java

File: platform/security/security-services-app/src/main/resources/features.xml [hunk 9]
```diff
@@ -127,6 +127,7 @@
 
     <feature name="security-sts-ldaplogin" install="manual" version="${project.version}"
              description="Security STS JAAS LDAP Login.">
+        <bundle>mvn:org.apache.commons/commons-pool2/2.5.0</bundle>
```
Comment (jlcsmith) [inline platform/security/security-services-app/src/main/resources/features.xml]: ❓ why is this needed?  Isn't commons-pool2 a simple jar not a bundle?

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/SslLdapLoginModule.java [hunk 10]
```diff
@@ -273,61 +262,26 @@ protected boolean doLogin() throws LoginException {
...
+          BindResult bindResult = connection.bind(userDn, tmpPassword);
 
-        if (!bindResult.isSuccess()) {
-          LOGGER.info("Bind failed");
+          if (!bindResult.isSuccess()) {
+            LOGGER.info("Bind failed");
+            return false;
+          }
+        } catch (Exception e) {
+          LOGGER.info("Unable to bind user to LDAP server.", e);
```
Comment (jlcsmith) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/SslLdapLoginModule.java]: ✏️ might be nice to include the userDn here

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/SslLdapLoginModule.java [hunk 11]
```diff
@@ -438,19 +386,32 @@ public void initialize(
...
+            serviceReferences
+                .stream()
+                .findFirst()
+                .orElseThrow(
+                    () ->
+                        new IllegalStateException(
+                            "No LDAPConnectionPool service found with id:" + connectionPoolId));
+        ldapConnectionPool = bundleContext.getService(serviceReference);
+      } catch (InvalidSyntaxException | IllegalStateException e) {
+        LOGGER.info("Unable to get LDAP Connection pool. LDAP log in will not be possible.", e);
```
Comment (jlcsmith) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/SslLdapLoginModule.java]: ✏️ this seems like a pretty severe issue.  Shouldn't this be logged at warn or even error?

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java (hunk 6 above)
Comment (ryeats) [reply platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java]: ill fix when I cherrypick to master

File: platform/security/security-services-app/src/main/resources/features.xml (hunk 9 above)
Comment (ryeats) [reply platform/security/security-services-app/src/main/resources/features.xml]: Its a bundle and a new dependency we don't directly use any version of commons-pool elsewhere in DDF. We use some libraries that do use the 1.6 commons-pool.

**Comment by ryeats** [comment]
build now

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java [hunk 12]
```diff
@@ -79,12 +103,104 @@
...
+      }
+      try {
+        LDAPConnectionFactory ldapConnectionFactory =
+            createLdapConnectionFactory(
+                (String) ldapProperties.get(LDAP_URL),
+                Boolean.parseBoolean((String) ldapProperties.get(START_TLS)));
+        ldapConnectionPool = new LDAPConnectionPool(ldapConnectionFactory, id);
+      } catch (LdapException e) {
+        LOGGER.error("Error creating ldap connection factory", e);
+      }
```
[the remark this answers is not in the document]
Comment (ryeats) [reply platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java]: copy and past issue I don't open a connection so the exception will never happen removing the try catch

**Comment by ryeats** [comment]
build now

**Comment by clockard** [comment]
build now
